File: problem_solving/intervals/cpp/56_merge_intervals.cpp
#include<vector>
#include<algorithm>
#include "../../utils/utils.h"
using namespace std;

vector<vector<int>> merge_intervals(vector<vector<int>> intervals) {
    if (intervals.empty()) {
        return intervals;
    }
    sort(intervals.begin(), intervals.end());
    vector<vector<int>> merged = {intervals[0]};
    for(int i = 1; i < intervals.size(); ++i ) {
        vector<int>& current = intervals[i]; 
        vector<int>& latest_merged = merged.back();
        if (current[0] >= latest_merged[0] && current[0] <= latest_merged[1] ) {
            latest_merged[1] = max(latest_merged[1], current[1]);
        } else {
            merged.push_back(current);
        }
    }
    return merged;
}




int main() {
    vector<vector<int>> intervals = {{1,3},{2,6},{8,10},{15,18}};
    vector<vector<int>> results = merge_intervals(intervals);
    //print_vector(result);
    for(const auto& r : results) {
        print_vector(r);
    }
    //Output: {{1,6},{8,10},{15,18}}


    intervals = {{1,4},{4,5}};
    results = merge_intervals(intervals);
    for(const auto& r : results) {
        print_vector(r);
    }
    //print_vector(result);
    //Output: {{1,5}}
    return 0;
}

File: problem_solving/intervals/py/56_merge_intervals.py
'''
https://leetcode.com/problems/merge-intervals/description/?envType=study-plan-v2&envId=top-interview-150

sort on start-key
create this interval... and if next start lies in this interval... then adjust the end-point of the interval
'''

#O(n * logn)  => for sorting
def merge_intervals(nums:list[list[int]])->list[list[int]]:
    if not nums:
        return []
    nums.sort(key=lambda t:t[0])
    result = [nums[0]]
    j = 0
    for i in range(1, len(nums)):
        if nums[i][0] >= result[j][0] and nums[i][0] <= result[j][1]:
            result[j] = [min(result[j][0], nums[i][0]), max(result[j][1], nums[i][1])]
        else:
            result.append(nums[i])
            j += 1
    return result


nums = [[1,3],[2,6],[15,18],[8,10]]
result = merge_intervals(nums)
print("result:" + str(result))


nums = [[2,6],[1,3],[8,10],[7,9]]
result = merge_intervals(nums)
print("result:" + str(result))